csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -469,7 +479,24 @@ public void WriteTo(CodedOutputStream output, Codec codec)
         [SecuritySafeCritical]
         public void WriteTo(ref WriteContext ctx, Codec codec)
         {
-            foreach (var entry in list)
+            if (ctx.state.CodedOutputStream is not null && ctx.state.CodedOutputStream.IsSerializationDeterministic())
+            {
+                // We can't sort the list in place, as that would invalidate the linked list.
+                // Instead, we create a new list, sort that, and then write it out.
+                var sorted = new List<KeyValuePair<TKey, TValue>>(list);
+                sorted.Sort((pair1, pair2) => Comparer<TKey>.Default.Compare(pair1.Key, pair2.Key));
```

Comment:
Given the repetition here, I'd probably extract the logic into a method of `GetSortedListCopy` or similar. Otherwise if we tweaked how the comparer is set up, it could end up inconsistent.

Reply:
refactored to `GetSortedListCopy`